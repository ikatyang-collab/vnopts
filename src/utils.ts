import {
  DefaultResult,
  DeprecatedResult,
  ForwardResult,
  InvalidHandler,
  NormalizedDefaultResult,
  NormalizedDeprecatedResult,
  NormalizedDeprecatedResultWithTrue,
  NormalizedForwardResult,
  NormalizedInvalidHandler,
  NormalizedRedirectResult,
  NormalizedTransferResult,
  NormalizedValidateResult,
  Options,
  RedirectResult,
  TransferResult,
  ValidateResult,
} from './types';

export type NotEmptyArray<T> = [T, ...T[]];

export type RecordFromUnion<T extends Options, K extends keyof T> = {
  [X in T[K]]: Extract<T, Record<K, X>>
};

export function recordFromArray<
  T extends Record<string, any>,
  K extends keyof T
>(array: T[], mainKey: K): RecordFromUnion<T, K> {
  const record = Object.create(null) as Partial<RecordFromUnion<T, K>>;

  for (const value of array) {
    const key = value[mainKey];

    // istanbul ignore next
    if (record[key]) {
      throw new Error(`Duplicate ${mainKey} ${JSON.stringify(key)}`);
    }

    // @ts-ignore
    record[key] = value;
  }

  return record as RecordFromUnion<T, K>;
}

export function mapFromArray<T extends Record<string, any>, K extends keyof T>(
  array: T[],
  mainKey: K,
): Map<T[K], T> {
  const map = new Map<T[K], T>();

  for (const value of array) {
    const key = value[mainKey];

    // istanbul ignore next
    if (map.has(key)) {
      throw new Error(`Duplicate ${mainKey} ${JSON.stringify(key)}`);
    }

    map.set(key, value);
  }

  return map;
}

export function createAutoChecklist() {
  const map: { [idString: string]: boolean } = Object.create(null);
  return (id: any) => {
    const idString = JSON.stringify(id);
    if (map[idString]) {
      return true;
    }
    map[idString] = true;
    return false;
  };
}

export function partition<T>(
  array: T[],
  predicate: (x: T) => boolean,
): [T[], T[]] {
  const trueArray: T[] = [];
  const falseArray: T[] = [];

  for (const value of array) {
    if (predicate(value)) {
      trueArray.push(value);
    } else {
      falseArray.push(value);
    }
  }

  return [trueArray, falseArray];
}

export function isInt(value: number) {
  return value === Math.floor(value);
}

export function comparePrimitive(
  a: undefined | null | boolean | number | string,
  b: undefined | null | boolean | number | string,
): number {
  if (a === b) {
    return 0;
  }

  const typeofA = typeof a;
  const typeofB = typeof b;

  const orders = [
    'undefined',
    'object', // null
    'boolean',
    'number',
    'string',
  ];
  if (typeofA !== typeofB) {
    return orders.indexOf(typeofA) - orders.indexOf(typeofB);
  }

  if (typeofA !== 'string') {
    return Number(a) - Number(b);
  }

  return (a as string).localeCompare(b as string);
}

export function normalizeInvalidHandler(
  invalidHandler: InvalidHandler,
): NormalizedInvalidHandler {
  return (...args) => {
    const errorMessageOrError = invalidHandler(...args);
    return typeof errorMessageOrError === 'string'
      ? new Error(errorMessageOrError)
      : /* istanbul ignore next*/ errorMessageOrError;
  };
}

export function normalizeDefaultResult<$Value>(
  result: DefaultResult<$Value>,
): NormalizedDefaultResult<$Value> {
  return result === undefined ? {} : result;
}

export function normalizeValidateResult(
  result: ValidateResult,
  value: unknown,
): NormalizedValidateResult {
  return result === true ? true : result === false ? { value } : result;
}

export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
): NormalizedDeprecatedResult<$Value>;
export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
  doNotNormalizeTrue: true,
): NormalizedDeprecatedResultWithTrue<$Value>;
export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
  doNotNormalizeTrue = false,
):
  | NormalizedDeprecatedResult<$Value>
  | NormalizedDeprecatedResultWithTrue<$Value> {
  return result === false
    ? false
    : result === true
      ? doNotNormalizeTrue
        ? true
        : [{ value }]
      : 'value' in result
        ? [result]
        : result.length === 0
          ? false
          : (result as NotEmptyArray<{ value: $Value }>);
}

export function normalizeTransferResult<$Value>(
  result: TransferResult<$Value>,
  value: $Value,
): NormalizedTransferResult<$Value> {
  return typeof result === 'string' || 'key' in result
    ? { from: value, to: result }
    : 'from' in result
      ? { from: result.from!, to: result.to }
      : { from: value, to: result.to };
}

export function normalizeForwardResult<$Value>(
  result: ForwardResult<$Value>,
  value: $Value,
): NormalizedForwardResult<$Value> {
  return result === undefined
    ? []
    : Array.isArray(result)
      ? result.map(transferResult =>
          normalizeTransferResult(transferResult, value),
        )
      : [normalizeTransferResult(result, value)];
}

export function normalizeRedirectResult<$Value>(
  result: RedirectResult<$Value>,
  value: $Value,
): NormalizedRedirectResult<$Value> {
  const redirect = normalizeForwardResult(
    typeof result === 'object' && 'redirect' in result
      ? result.redirect
      : result,
    value,
  );
  return redirect.length === 0
    ? { remain: value, redirect }
    : typeof result === 'object' && 'remain' in result
      ? { remain: result.remain, redirect }
      : { redirect };
}
