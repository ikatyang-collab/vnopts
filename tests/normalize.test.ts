import { describe, expect, test } from 'vitest'
import {
  createSchema,
  Normalizer,
  OptionKey,
  OptionValue,
  Schema,
  SchemaParameters,
} from '../src/index.js'
import { createLogger, eachHandler } from './__helpers__/utils.js'

class TestSchema extends Schema<any, SchemaParameters<any>> {
  public expected() {
    return '<expected>'
  }
  public validate() {
    return true
  }
}

const name = '<key>'
const value = '<value>'
const option = { [name]: value }
const transferKey = '<transfer-key>'
const transferValue = '<transfer-value>'
const transferPair = { key: transferKey, value: transferValue }
const transfer = { [transferKey]: transferValue }
const unknownKey = '<unknown-key>'
const unknownValue = '<unknown-value>'
const unknown = { [unknownKey]: unknownValue }
const veryUnknown = { veryUnknown: '<very-unknown-value>' }
const overlapKey = '<overlap-key>'
const overlapValue = '<overlap-value>'
const overlap = { [overlapKey]: overlapValue }
const invalidValue = '<invalid-value>'
const knownKey = '<known-key>'

const passThrough = (k: OptionKey, v: OptionValue) => ({ [k]: v })
const redirectTo = (k: OptionKey) => (_k: OptionKey, v: OptionValue) => ({
  [k]: v,
})

describe.each`
  parameters                                           | input                        | output                                 | hasWarnings
  ${{ default: { value } }}                            | ${{}}                        | ${option}                              | ${false}
  ${{ validate: false }}                               | ${option}                    | ${Error}                               | ${false}
  ${{ validate: { value: invalidValue } }}             | ${option}                    | ${Error}                               | ${false}
  ${{ deprecated: true }}                              | ${option}                    | ${option}                              | ${true}
  ${{ deprecated: { value } }}                         | ${option}                    | ${option}                              | ${true}
  ${{ deprecated: true, redirect: transferKey }}       | ${option}                    | ${{ [transferKey]: value }}            | ${true}
  ${{ deprecated: { value }, redirect: transferKey }}  | ${option}                    | ${{ [transferKey]: value }}            | ${true}
  ${{ deprecated: true, redirect: transferPair }}      | ${option}                    | ${{ ...transfer }}                     | ${true}
  ${{ deprecated: { value }, redirect: transferPair }} | ${option}                    | ${{ ...transfer }}                     | ${true}
  ${{ forward: transferKey }}                          | ${option}                    | ${{ ...option, [transferKey]: value }} | ${false}
  ${{ forward: transferPair }}                         | ${option}                    | ${{ ...option, ...transfer }}          | ${false}
  ${{ redirect: transferKey }}                         | ${option}                    | ${{ [transferKey]: value }}            | ${false}
  ${{ redirect: transferPair }}                        | ${option}                    | ${{ ...transfer }}                     | ${false}
  ${{}}                                                | ${unknown}                   | ${{}}                                  | ${true}
  ${{}}                                                | ${veryUnknown}               | ${{}}                                  | ${true}
  ${{ unknown: passThrough }}                          | ${unknown}                   | ${unknown}                             | ${false}
  ${{ unknown: redirectTo(knownKey) }}                 | ${unknown}                   | ${{ [knownKey]: unknownValue }}        | ${false}
  ${{ redirect: overlapKey }}                          | ${{ ...option, ...overlap }} | ${{ ...overlap }}                      | ${false}
`(
  '',
  eachHandler<TestSchema>(
    TestSchema,
    { name },
    {
      schemas: [
        createSchema(TestSchema, { name: knownKey }),
        createSchema(TestSchema, { name: overlapKey }),
        createSchema(TestSchema, { name: transferKey }),
      ],
    },
  ),
)

test('no duplicate deprecation warnings', () => {
  for (const deprecatedResult of [true, { value }]) {
    const logger = createLogger()

    const schemas = [
      TestSchema.create<TestSchema>({ name, deprecated: deprecatedResult }),
    ]
    const normalizer = new Normalizer(schemas, { logger })

    let lastMessagesLength = -1
    for (let i = 0; i < 2; i++) {
      normalizer.normalize({ [name]: value })
      const messagesLength = logger.getMessages().length

      if (i > 0) {
        expect(messagesLength).toEqual(lastMessagesLength)
        expect(messagesLength).toBeGreaterThan(0)
      }

      lastMessagesLength = messagesLength
    }
  }
})
