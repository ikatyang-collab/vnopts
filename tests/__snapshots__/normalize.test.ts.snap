// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema {"name":"<key>","deprecated":{"value":"<value>"},"redirect":"<transfer-key>"} with input {"<key>":"<value>"} 1`] = `
Array [
  "{ \\"<key>\\": \\"<value>\\" } is deprecated; we now treat it as { \\"<transfer-key>\\": \\"<value>\\" }.",
]
`;

exports[`schema {"name":"<key>","deprecated":{"value":"<value>"},"redirect":{"key":"<transfer-key>","value":"<transfer-value>"}} with input {"<key>":"<value>"} 1`] = `
Array [
  "{ \\"<key>\\": \\"<value>\\" } is deprecated; we now treat it as { \\"<transfer-key>\\": \\"<transfer-value>\\" }.",
]
`;

exports[`schema {"name":"<key>","deprecated":{"value":"<value>"}} with input {"<key>":"<value>"} 1`] = `
Array [
  "{ \\"<key>\\": \\"<value>\\" } is deprecated.",
]
`;

exports[`schema {"name":"<key>","deprecated":true,"redirect":"<transfer-key>"} with input {"<key>":"<value>"} 1`] = `
Array [
  "\\"<key>\\" is deprecated; we now treat it as \\"<transfer-key>\\".",
]
`;

exports[`schema {"name":"<key>","deprecated":true,"redirect":{"key":"<transfer-key>","value":"<transfer-value>"}} with input {"<key>":"<value>"} 1`] = `
Array [
  "\\"<key>\\" is deprecated; we now treat it as { \\"<transfer-key>\\": \\"<transfer-value>\\" }.",
]
`;

exports[`schema {"name":"<key>","deprecated":true} with input {"<key>":"<value>"} 1`] = `
Array [
  "\\"<key>\\" is deprecated.",
]
`;

exports[`schema {"name":"<key>","validate":{"value":"<invalid-value>"}} with input {"<key>":"<value>"} 1`] = `"Invalid \\"<key>\\" value. Expected <expected>, but received \\"<invalid-value>\\"."`;

exports[`schema {"name":"<key>","validate":false} with input {"<key>":"<value>"} 1`] = `"Invalid \\"<key>\\" value. Expected <expected>, but received \\"<value>\\"."`;

exports[`schema {"name":"<key>"} with input {"<unknown-key>":"<unknown-value>"} 1`] = `
Array [
  "Ignored unknown option { \\"<unknown-key>\\": \\"<unknown-value>\\" }. Did you mean \\"<known-key>\\"?",
]
`;

exports[`schema {"name":"<key>"} with input {"veryUnknown":"<very-unknown-value>"} 1`] = `
Array [
  "Ignored unknown option { veryUnknown: \\"<very-unknown-value>\\" }.",
]
`;
